kf/bpf.go

```diff
@@ -522,137 +525,182 @@ func (b *BPF) VerifyAndGetArtifacts(conf *config.Config) error {
-					extractedFilePath,
-					os.O_WRONLY|os.O_CREATE|os.O_TRUNC,
-					file.Mode(),
-				)
+			tempDir := filepath.Join(conf.BPFDir, b.Program.Name, b.Program.Version)
+
+			for _, file := range zipReader.File {
+
+				zippedFile, err := file.Open()
 				if err != nil {
-					return fmt.Errorf("unzip failed to create file: %w", err)
+					return fmt.Errorf("unzip failed: %w", err)
 				}
-				defer outputFile.Close()
+				defer zippedFile.Close()
 
-				buf := copyBufPool.Get().(*bytes.Buffer)
-				_, err = io.CopyBuffer(outputFile, zippedFile, buf.Bytes())
-				if err != nil {
-					return fmt.Errorf("GetArtifacts failed to copy files: %w", err)
+				extractedFilePath := filepath.Join(
+					tempDir,
+					file.Name,
+				)
+				if !strings.HasPrefix(extractedFilePath, filepath.Clean(tempDir)+string(os.PathSeparator)) {
+					return fmt.Errorf("invalid file path: %s", extractedFilePath)
+				}
+				if file.FileInfo().IsDir() {
+					os.MkdirAll(extractedFilePath, file.Mode())
+				} else {
+					outputFile, err := os.OpenFile(
+						extractedFilePath,
+						os.O_WRONLY|os.O_CREATE|os.O_TRUNC,
+						file.Mode(),
+					)
+					if err != nil {
+						return fmt.Errorf("unzip failed to create file: %w", err)
+					}
+					defer outputFile.Close()
+
+					buf := copyBufPool.Get().(*bytes.Buffer)
+					_, err = io.CopyBuffer(outputFile, zippedFile, buf.Bytes())
+					if err != nil {
+						return fmt.Errorf("GetArtifacts failed to copy files: %w", err)
+					}
+					copyBufPool.Put(buf)
 				}
-				copyBufPool.Put(buf)
 			}
+			newDir := strings.Split(b.Program.Artifact, ".")
+			b.FilePath = filepath.Join(tempDir, newDir[0])
+			return nil
 		}
-		newDir := strings.Split(b.Program.Artifact, ".")
-		b.FilePath = filepath.Join(tempDir, newDir[0])
-		return nil
-	} else if strings.HasSuffix(b.Program.Artifact, ".tar.gz") {
-		archive, err := gzip.NewReader(buf)
-		if err != nil {
-			return fmt.Errorf("failed to create Gzip reader: %w", err)
-		}
-		defer archive.Close()
-		tarReader := tar.NewReader(archive)
-		tempDir := filepath.Join(conf.BPFDir, b.Program.Name, b.Program.Version)
+	case strings.HasSuffix(b.Program.Artifact, ".tar.gz"):
+		{
+			archive, err := gzip.NewReader(buf)
+			if err != nil {
+				return fmt.Errorf("failed to create Gzip reader: %w", err)
+			}
+			defer archive.Close()
+			tarReader := tar.NewReader(archive)
+			tempDir := filepath.Join(conf.BPFDir, b.Program.Name, b.Program.Version)
 
-		for {
-			header, err := tarReader.Next()
+			for {
+				header, err := tarReader.Next()
 
-			if err == io.EOF {
-				break
-			} else if err != nil {
-				return fmt.Errorf("untar failed: %w", err)
-			}
+				if err == io.EOF {
+					break
+				} else if err != nil {
+					return fmt.Errorf("untar failed: %w", err)
+				}
 
-			if strings.Contains(header.Name, "..") {
-				return fmt.Errorf("zipped file contians filepath (%s) that includes (..)", header.Name)
-			}
+				if strings.Contains(header.Name, "..") {
```

Comment:
Can we have a similar check as the zip filename check above with filepath.Clean() ?  It probably also makes breaking out this code, so as we add support for new compression algorithms the filename checking logic remains the same and can be managed in single place.

Reply:
if we check for somewhere else using the helper function it is giving security issues.

Reply:
breaking out into a separate function vs. having similar code 2+ more times seems to make sense, not sure why this is alerting if we make this change.  We can discuss more on call